File: C_plus_plus/Reverse_Linked_List_II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *reverseBetween(ListNode *head, int m, int n) {
        ListNode *p = head,*last = NULL,*begin = NULL,*end = NULL,*before = NULL,*a = NULL,*b=NULL;
        int pos = 1;
        bool reverseFlag = false;
        while(p){
            if(pos == m){
                begin = last;
                a = p;
                reverseFlag = true;
            }
            if(pos == n){
                end = p->next;
                b = p;
                p->next = before;
                break;
            }
            if(reverseFlag){
                ListNode * temp = p;
                p = p->next;
                temp->next = before;
                before = temp;
            }
            else{
                last = p;
                p = p->next;
            }
            ++pos;
        }
        if(begin)
            begin->next = b;
        if(a)
            a->next = end;
        if(m==1) return b;
        else return head;
    }
};

File: java/Reverse_Linked_List_II.java
// First, locate the m node, and then reverse the n-m+1 nodes, 
// in which process we should record the previous last one of unreversed nodes
// and the after first one of unreversed nodes.
// And then , recombination the three parts.
// One pass, time is O(n), space is O(1).

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        int count = 0;
        ListNode pre = null, after = null, p = head, last = null, tail = null;
        while(p != null) {
            if (++count == m)
                break;
            pre = p;
            p = p.next;
        }
        for (int i = 0; i < n-m+1; ++i) {
            ListNode temp = p.next;
            if (last == null) {
                tail = last = p;
            }
            else {
                p.next = last;
                last = p;
            }
            p = temp;
        }
        if (pre == null) head = last;
        else pre.next = last;
        tail.next = p;
        return head;
    }
}
